### PR TITLE
add const generics blog post

### PR DESCRIPTION
**Pull request by lcnr**
first draft written by @varkor and me.

cc @rust-lang/project-const-generics @rust-lang/lang

### CODE REVIEW

**Review by oli-obk — APPROVED** [review]

**Comment by nikomatsakis** [comment]
cc @rust-lang/core -- this looks good to me, perhaps folks want to take a look? It'd be good to get it merged soon.

File: posts/2021-02-11-const-generics-mvp-beta.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+#### Only integral types are permitted for const generics
+
+For now, the only types that may be used as the type of a const generic argument are the types of integers (i.e. signed and unsigned integers, including `isize` and `usize`) as well as `char` and `bool`. This covers a primary use case of const, namely abstracting over arrays. In the future, this restriction will be lifted to allow more complex types, such as `&str` and user-defined types.
+
+#### No complex generic expressions in const arguments
+
+Currently, const parameters may only be instantiated by const arguments of the current forms:
```
Comment (camelid) [inline posts/2021-02-11-const-generics-mvp-beta.md:40]: What does "of the current form" mean? Maybe you meant to remove that in favor of "Currently,"?

Comment (jyn514) [reply posts/2021-02-11-const-generics-mvp-beta.md:40]: I think they might have meant "of the following forms".

Comment (camelid) [reply posts/2021-02-11-const-generics-mvp-beta.md:40]: > I think they might have meant "of the following forms".

Ah, that's much clearer. Can it be changed to that?

Comment (lcnr) [reply posts/2021-02-11-const-generics-mvp-beta.md:40]:
```suggestion
Currently, const parameters may only be instantiated by const arguments of the following forms:
```

**Comment by Mark-Simulacrum** [comment]
This looks good to me -- I'm thinking maybe let's merge this Wednesday (the 24th), unless some more problems are found? I'd be happy to do that if @lcnr you wanted to update the date on the post.

**Comment by lcnr** [comment]
seems good to me

File: posts/2021-02-11-const-generics-mvp-beta.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ author: The const generics project group
 
 After more than 3 years since the [original RFC for const generics](https://github.com/rust-lang/rfcs/blob/master/text/2000-const-generics.md) was accepted, **the first version of const generics is now available in the Rust beta channel!** It will be available in the 1.51 release, which is expected to be released on **March 25th, 2021**. Const generics is one of the [most highly anticipated](https://blog.rust-lang.org/2020/12/16/rust-survey-2020.html) features coming to Rust, and we're excited for people to start taking advantage of the increased power of the language following this addition.
 
-Even if you don't know what const generics are (in which case, read on!), you've likely been benefitting from them: const generics are already employed in the Rust compiler to improve the ergonomics of arrays and diagnostics; more on that below.
+Even if you don't know what const generics are (in which case, read on!), you've likely been benefitting from them: const generics are already employed in the Rust standard library to improve the ergonomics of arrays and diagnostics; more on that below.
```
Comment (camelid) [inline posts/2021-02-11-const-generics-mvp-beta.md]: Typo:

```suggestion
Even if you don't know what const generics are (in which case, read on!), you've likely been benefiting from them: const generics are already employed in the Rust standard library to improve the ergonomics of arrays and diagnostics; more on that below.
```

Comment (camelid) [reply posts/2021-02-11-const-generics-mvp-beta.md]: Actually it looks like either spelling is fine, so go with whichever you prefer :)

**Comment by lcnr** [comment]
@Mark-Simulacrum can we go ahead and merge this?

And if so, should I update the date of the post again, or can you quickly do that?
Afaik the only required change is the file name

**Comment by Mark-Simulacrum** [comment]
Yeah, sorry, forgot about this - will bump date and merge later today.

**Review by Mark-Simulacrum — APPROVED** [review]

File: posts/2021-02-26-const-generics-mvp-beta.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+There are several complexities involved in supporting complex expressions. A feature flag, `feature(const_evaluatable_checked)`, is available in the Nightly channel, which enables a version of complex expression support for const generics.
+
+One difficulty lies in the necessity of having some way to compare unevaluated constants, as the compiler does not automatically know that two syntactically identical expressions are actually equal. This involves a kind of symbolic reasoning about expressions, which is a complex problem in general.
+```rust
+// The two expressions `N + 1` and `N + 1` are distinct
+// entities in the compiler, so we need a way to check
+// if they should be considered equal.
+fn foo<const N: usize>() -> [u8; N + 1] {
+    [u8; N + 1]
```
Comment (varkor) [inline posts/2021-02-26-const-generics-mvp-beta.md:105]: This should be `[0; N + 1]` or similar.

Comment (jyn514) [reply posts/2021-02-26-const-generics-mvp-beta.md:105]: This was fixed in https://github.com/rust-lang/blog.rust-lang.org/pull/785.

Comment (varkor) [reply posts/2021-02-26-const-generics-mvp-beta.md:105]: Ah, perfect :)

**Pull request merged**